### DISCUSSION TITLE
How do I index my data according to the years

### DISCUSSION THREAD

**Discussion by delahoussou** [pydata/xarray#7412] [Q&A]
I have this dataset. I want to calculate the wind speed for each year. The data is from 2000 to 2013. I can calculate the wind speed. 
Is there any a more efficient way that I can index my dataset. The code I have right now I have to rewrite the indexing part for each indvidivual year manual. Is there a way better way I can do the indexing and break the data into individual years? 

# opens all the netcdf4 files containing eu10 and ev10

#present climate data
present_eu10 = xr.open_mfdataset('/chinook2/wgallus/climate_output/*EU10*.nc')
present_ev10 = xr.open_mfdataset('/chinook2/wgallus/climate_output/*EV10*.nc')

#future climate data
future_eu10 = xr.open_mfdataset('/chinook2/wgallus/climate_output_future/*PGW_EU10*.nc')
future_ev10= xr.open_mfdataset('/chinook2/wgallus/climate_output_future/*PGW_EV10*.nc')

# Indexing the data according to the different years. 

year2000eu = present_eu10['EU10'].isel(Time = slice(0, 2208))
year2000ev = present_ev10['EV10'].isel(Time = slice(0, 2208))
future2000eu = future_eu10['EU10'].isel(Time = slice(0, 2208))
future2000ev = future_ev10['EV10'].isel(Time = slice(0, 2208))

ds1_2001 = present_eu10['EU10'].isel(Time = slice(2208, 10968))
ds2_2001 = present_ev10['EV10'].isel(Time = slice(2208, 10968))
ds3_2001 = future_eu10['EU10'].isel(Time = slice(2208, 10968))
ds4_2001 = future_ev10['EV10'].isel(Time = slice(2208, 10968))

ds1_2002 = present_eu10['EU10'].isel(Time = slice(10968, 19728))
ds2_2002 = present_ev10['EV10'].isel(Time = slice(10968, 19728))
ds3_2002 = future_eu10['EU10'].isel(Time = slice( 10968, 19728))
ds4_2002 = future_ev10['EV10'].isel(Time = slice( 10968, 19728))

ds1_2003 = present_eu10['EU10'].isel(Time = slice(19728, 28488))
ds2_2003 = present_ev10['EV10'].isel(Time = slice(19728, 28488))
ds3_2003 = future_eu10['EU10'].isel(Time = slice(19728,28488))
ds4_2003 = future_ev10['EV10'].isel(Time = slice(19728,28488))

ds1_2004 = present_eu10['EU10'].isel(Time = slice(28488, 37272))
ds2_2004 = present_ev10['EV10'].isel(Time = slice(28488, 37272))
ds3_2004 = future_eu10['EU10'].isel(Time = slice(28488, 37272))
ds4_2004 = future_ev10['EV10'].isel(Time = slice(28488, 37272))

ds1_2005 = present_eu10['EU10'].sel(Time = "2005")
ds2_2005 = present_ev10['EV10'].sel(Time = "2005")
ds3_2005 = future_eu10['EU10'].sel(Time = "2005")
ds4_2005 = future_ev10['EV10'].sel(Time = "2005")

ds1_2006 = present_eu10['EU10'].sel(Time = "2006")
ds2_2006 = present_ev10['EV10'].sel(Time = "2006")
ds3_2006 = future_eu10['EU10'].sel(Time = "2006")
ds4_2006 = future_ev10['EV10'].sel(Time = "2006")

ds1_2007 = present_eu10['EU10'].sel(Time = "2007")
ds2_2007 = present_ev10['EV10'].sel(Time = "2007")
ds3_2007 = future_eu10['EU10'].sel(Time =  "2007")
ds4_2007 = future_ev10['EV10'].sel(Time =  "2007")

ds1_2008 = present_eu10['EU10'].sel(Time = "2008")
ds2_2008 = present_ev10['EV10'].sel(Time = "2008")
ds3_2008 = future_eu10['EU10'].sel(Time = "2008")
ds4_2008 = future_ev10['EV10'].sel(Time = "2008")

ds1_2009 = present_eu10['EU10'].sel(Time = "2009")
ds2_2009 = present_ev10['EV10'].sel(Time = "2009")
ds3_2009 = future_eu10['EU10'].sel(Time = "2009")
ds4_2009 = future_ev10['EV10'].sel(Time = "2009")

ds1_2010 = present_eu10['EU10'].sel(Time = "2010")
ds2_2010 = present_ev10['EV10'].sel(Time = "2010")
ds3_2010 = future_eu10['EU10'].sel(Time = "2010")
ds4_2010 = future_ev10['EV10'].sel(Time = "2010")

ds1_2011 = present_eu10['EU10'].sel(Time = "2011")
ds2_2011 = present_ev10['EV10'].sel(Time = "2011")
ds3_2011 = future_eu10['EU10'].sel(Time = "2011")
ds4_2011 = future_ev10['EV10'].sel(Time = "2011")

ds1_2012 = present_eu10['EU10'].sel(Time = "2012")
ds2_2012 = present_ev10['EV10'].sel(Time = "2012")
ds3_2012 = future_eu10['EU10'].sel(Time = "2012")
ds4_2012 = future_ev10['EV10'].sel(Time = "2012")

ds1_2013 = present_eu10['EU10'].sel(Time = "2013")
ds2_2013 = present_ev10['EV10'].sel(Time = "2013")
ds3_2013 = future_eu10['EU10'].sel(Time = "2013")
ds4_2013 = future_ev10['EV10'].sel(Time = "2013")

**Comment by keewis** [comment]
this sounds like a [`groupby`](https://docs.xarray.dev/en/stable/user-guide/groupby.html) operation:
```python
present_eu10.groupby("Time.year")
```

Does that help? If not, can you give a bit more context on the datasets and what you're doing with the extracted groups?

**Reply by delahoussou** [reply]
Thank you for the answer. What I am trying to do is this. I have this data and I want to calculate the wind speed and get the maximum wind speed for each year. Here is a picture of what is in the folder. I opened all the files containing eu10 and ev10. To calculate the speed is : (eu10 **2 + ev10**2) ** 0.5 . "groupby"  allows me to group them by year. But then I cannot calculate the speed. I get a TypeError
I hope the folliowing will give you more context 


# opens all the netcdf4 files containing eu10 and ev10
#present climate data
present_eu10 = xr.open_mfdataset('/chinook2/wgallus/climate_output/*EU10*.nc').groupby("Time.year")
present_ev10 = xr.open_mfdataset('/chinook2/wgallus/climate_output/*EV10*.nc').groupby("Time.year")
#future climate data
future_eu10 = xr.open_mfdataset('/chinook2/wgallus/climate_output_future/*PGW_EU10*.nc').groupby("Time.year")
future_ev10= xr.open_mfdataset('/chinook2/wgallus/climate_output_future/*PGW_EV10*.nc').groupby("Time.year")

speed = (present_eu10 **2 + present_ev10 ** 2) ** 0.5
TypeError                                 Traceback (most recent call last)
Cell In[29], line 1
----> 1 speed = (present_eu10 **2 + present_ev10 ** 2) ** 0.5

TypeError: GroupBy objects only support binary ops when the other argument is a Dataset or DataArray

This is what in my folder. I open only the files containing eu10 and ev10 
for data in (glob.glob('/chinook2/wgallus/climate_output_future/*.nc')):
    print(data)

/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_EU10_200010-200012.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_EV10_200010-200012.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_PREC_ACC_NC_200010-200012.nc
/chinook2/wgallus/climate_output_future/wrf3d_d01_PGW_QRAIN_200010.nc
/chinook2/wgallus/climate_output_future/wrf3d_d01_PGW_QRAIN_200011.nc
/chinook2/wgallus/climate_output_future/wrf3d_d01_PGW_QRAIN_200012.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_T2_200010-200012.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_EU10_200101-200103.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_EV10_200101-200103.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_PREC_ACC_NC_200101-200103.nc
/chinook2/wgallus/climate_output_future/wrf3d_d01_PGW_QRAIN_200101.nc
/chinook2/wgallus/climate_output_future/wrf3d_d01_PGW_QRAIN_200102.nc
/chinook2/wgallus/climate_output_future/wrf3d_d01_PGW_QRAIN_200103.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_T2_200101-200103.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_EU10_200104-200106.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_EV10_200104-200106.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_PREC_ACC_NC_200104-200106.nc
/chinook2/wgallus/climate_output_future/wrf3d_d01_PGW_QRAIN_200104.nc
/chinook2/wgallus/climate_output_future/wrf3d_d01_PGW_QRAIN_200105.nc
/chinook2/wgallus/climate_output_future/wrf3d_d01_PGW_QRAIN_200106.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_T2_200104-200106.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_EU10_200107-200109.nc
/chinook2/wgallus/climate_output_future/wrf2d_d01_PGW_EV10_200107-200109.nc

**Comment by DanielAdriaansen** [comment]
Can you provide the output from:
`print(present_eu10)`

and

`print(present_eu10['Time'])`

**Reply by delahoussou** [reply]
Yes. Kindly give me a few seconds

**Reply by delahoussou** [reply]
print(present_eu10)
<xarray.Dataset>
Dimensions:   (Time: 113952, south_north: 1015, west_east: 1359)
Coordinates:
  * Time      (Time) datetime64[ns] 2000-10-01 ... 2013-09-30T23:00:00
    XLAT      (south_north, west_east) float32 dask.array<chunksize=(1015, 1359), meta=np.ndarray>
    XLONG     (south_north, west_east) float32 dask.array<chunksize=(1015, 1359), meta=np.ndarray>
Dimensions without coordinates: south_north, west_east
Data variables:
    COSALPHA  (Time, south_north, west_east) float32 dask.array<chunksize=(2208, 1015, 1359), meta=np.ndarray>
    EU10      (Time, south_north, west_east) float32 dask.array<chunksize=(2208, 1015, 1359), meta=np.ndarray>
    SINALPHA  (Time, south_north, west_east) float32 dask.array<chunksize=(2208, 1015, 1359), meta=np.ndarray>
    Times     (Time) |S19 dask.array<chunksize=(2208,), meta=np.ndarray>
    XTIME     (Time) float32 dask.array<chunksize=(2208,), meta=np.ndarray>
Attributes: (12/113)
    NCO:                             "4.5.5"
    history:                         Tue Jan 17 13:08:43 2017: ncks -O -4 /gl...
    Division:                        NCAR/RAL/HAP
    Project:                         CONUS-HYDRO
    FileGenerated:                   20140813
    Source_Code:                     make_HYDRO_CONUS_analysis_files.scr
    ...                              ...
    SOUTH-NORTH_GRID_DIMENSION:      1016
    WEST-EAST_GRID_DIMENSION:        1360
    SIMULATION_START_DATE:           2000-10-01_00:00:00
    Publication_DOI:                 doi:10.1007/s00382-016-3327-9
    Conventions:                     CF-1.6
    TITLE:                            OUTPUT FROM WRF V3.4.1 MODEL , CURRENT ...

**Reply by delahoussou** [reply]
print(present_eu10['Time'])

<xarray.DataArray 'Time' (Time: 113952)>
array(['2000-10-01T00:00:00.000000000', '2000-10-01T01:00:00.000000000',
       '2000-10-01T02:00:00.000000000', ..., '2013-09-30T21:00:00.000000000',
       '2013-09-30T22:00:00.000000000', '2013-09-30T23:00:00.000000000'],
      dtype='datetime64[ns]')
Coordinates:
  * Time     (Time) datetime64[ns] 2000-10-01 ... 2013-09-30T23:00:00
Attributes:
    long_name:    Time
    description:  Time

**Reply by DanielAdriaansen** [reply]
I would review this section of the documentation: https://docs.xarray.dev/en/stable/user-guide/time-series.html#datetime-indexing, and then write a loop to loop in time so each iteration of your loop represents one year. Then you can compute the wind speed for that year by using the datetime indexing described in the documentation linked above.

What do you want to do after computing the wind speed?

**Reply by delahoussou** [reply]
I have being considering doing a for loop but I have a feeling that my loop will have about 13 if and elif statement. 
sel() I cannot use inside the for loop because it is index by label, with isel() the problem is each time it iterate the preceding index must become the starting index for the next iteration and so on.

**Reply by delahoussou** [reply]
I will give another try to a for loop with conditional statements.

**Reply by delahoussou** [reply]
After I calculate the wind speed this is what I wanna do: 
I wanna count how many times each year (these may be 3 hours periods) that the wind speed is is greater that 15 ms-1 and 20 ms-1 and keep track of peak value for each year.

**Reply by DanielAdriaansen** [reply]
Instead of a for loop you need to loop over actual times using the Python datetime module:
https://docs.python.org/3/library/datetime.html

```
import datetime

date_begin = datetime.datetime(2000,1,1,0,0,0)
date_end = datetime.datetime(2013,1,1,0,0,0)

while date_begin <= date_end:

  # Get the current year as a string
  current_year = date_begin.strftime("%Y")

  # Calculate the speed for this year
  speed = (present_eu10.sel(Time=current_year) **2 + present_ev10.sel(Time=current_year) ** 2) ** 0.5

  # Move to the next year
  date_begin = date_begin + datetime.timedelta(years=1)
```

**Reply by delahoussou** [reply]
>>> date_begin = datetime.datetime(2000,1,1,0,0,0)
       date_end = datetime.datetime(2013,1,1,0,0,0)
       while date_begin <= date_end:
        current_year = date_begin.strftime("%Y")
        speed = (present_eu10.sel(Time = current_year) ** 2 + present_ev10.sel(Time = current_year) ** 2) ** 0.5 
        date_begin = date_begin + datetime.timedelta(years=1)
   
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[16], line 6
      3 while date_begin <= date_end:
      4     current_year = date_begin.strftime("%Y")
----> 6     speed = (present_eu10.sel(Time = current_year) ** 2 + present_ev10.sel(Time = current_year) ** 2) ** 0.5 
      7     date_begin = date_begin + datetime.timedelta(years=1)

TypeError: unsupported operand type(s) for ** or pow(): 'Array' and 'int'

Also after playing with your code, I realized the last line throws an error 
  date_begin = date_begin + datetime.timedelta(years=1)
years is not defined. 

How do I address these two issues? Thank. you

**Reply by DanielAdriaansen** [reply]
Hi,

I thought that `datetime` had a `timedelta` object of "years" but I was wrong:
https://docs.python.org/3/library/datetime.html#timedelta-objects

You'll somehow need to increment time forward the duration you want, since if you use days=365, I don't think you'll account for leap years.

Perhaps a better approach is to create a list of years you'd like ahead of time like:
`years = [str(x) for x in range(2000,2014,1)]`

And then do a for loop over each year in that list instead of the while loop I suggested like:
```
years = [str(x) for x in range(2000,2014,1)]
for current_year in years:

  # Calculate the wind speed
  speed = (present_eu10.sel(Time=current_year) **2 + present_ev10.sel(Time=current_year) ** 2) ** 0.5
```

Or, the suggestion below from @keewis is also very good. Ultimately there are several approaches to accomplish what you want to do I believe.

**Reply by delahoussou** [reply]
Thank you for the suggestion both of them gave me errors. 
With this one in particular is it saying that I am trying to perform an operation on something that is not an Array or integer
speed = (present_eu10.sel(Time=current_year) **2 + present_ev10.sel(Time=current_year) ** 2) ** 0.5

**Reply by delahoussou** [reply]
I don't know why but I think sel() and isel() both return Array but in this case that is the error I get

**Reply by keewis** [reply]
yeah, this is pretty weird. Can you tell us the output of
```python
print(present_eu10["EU10"].data)
```
?

All arrays that `xarray` can wrap *should* support `__pow__` with integers.

It would also be useful to know what your environment is. Could you tell us the output of `xr.show_versions()`?

**Reply by delahoussou** [reply]
print(present_eu10["EU10"].data)
dask.array<concatenate, shape=(113952, 1015, 1359), dtype=float32, chunksize=(2208, 1015, 1359), chunktype=numpy.ndarray>

This is what I use
https://jupyterhub.agron.iastate.edu/

**Comment by keewis** [comment]
If I understand correctly, what you want to do is
1. compute the magnitude of the wind speed from wind vectors
2. get the maximum wind speed per year
3. compute how many times per year the magnitude exceeds a certain threshold

Breaking the task up this way makes the code very easy to read and understand:
```python
wind_speed = np.sqrt(present_eu10["EU10"] ** 2 + present_ev["EV10"] ** 2)
maximum_per_year = wind_speed.groupby("Time.year").max()
n_exceeded = (wind_speed > threshold).groupby("Time.year").sum()
```
if `threshold` is something like `DataArray([10, 15, 20], dims="thresholds")`, this will allow you to compute all thresholds in a single operation. If you're instead interested in binning the wind speeds: that is possible as well, but unfortunately somewhat tricky at the moment (we're working on that, though).

**Reply by delahoussou** [reply]
This is the code:  
treshold = xr.DataArray([10,15], dims = "thresholds")
wind_speed_magnitude = np.sqrt (present_eu10["EU10"] ** 2 + present_ev10["EV10"] ** 2) ** 0.5 
maximum_per_year = wind_speed_magnitude.groupby("Time.year").max()
n_exceeded = (wind_speed_magnitude > treshold).groupby("Time.year").sum()

error: 
To avoid creating the large chunks, set the option
    >>> with dask.config.set(**{'array.slicing.split_large_chunks': True}):
    ...     array[indexer]
  return self.array[key]

Then I did this: 
 with dask.config.set(**{'array.slicing.split_large_chunks': True}):
        treshold = xr.DataArray([10,15], dims = "thresholds")
        wind_speed_magnitude = np.sqrt (present_eu10["EU10"] ** 2 + present_ev10["EV10"] ** 2) ** 0.5 
        maximum_per_year = wind_speed_magnitude.groupby("Time.year").max()
        n_exceeded = (wind_speed_magnitude > treshold).groupby("Time.year").sum()

output: 
   n_exceeded
xarray.DataArrayyear: 14south_north: 1015west_east: 1359thresholds: 2
Coordinates:
XLAT
(south_north, west_east)
float32
18.14 18.15 18.15 ... 52.43 52.42
XLONG
(south_north, west_east)
float32
-122.9 -122.8 ... -57.0 -56.95
year
(year)
int64
2000 2001 2002 ... 2011 2012 2013
array([2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011,
       2012, 2013])
Indexes:
year
PandasIndex
PandasIndex(Int64Index([2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
            2011, 2012, 2013],
           dtype='int64', name='year'))

**Reply by keewis** [reply]
that seems like it worked.

For the future: text-copying the HTML repr makes it hard to extract the data, so it is better to either post screenshots of the HTML repr or use `xr.set_options(display_style="text")` to get the text representation.

**Reply by delahoussou** [reply]
How do I retrieve the data that is: 
get the maximum wind speed per year
compute how many times per year the magnitude exceeds a certain threshold

**Reply by keewis** [reply]
you mean, because the data is still a `dask` array, so the computation didn't actually happen, yet? Just call `.compute()` on it. What I would do is collect everything in a `Dataset` and compute everything together:

```python
wind_speed = np.sqrt(present_eu10["EU10"] ** 2 + present_ev["EV10"] ** 2)
maximum_per_year = wind_speed.groupby("Time.year").max()
n_exceeded = (wind_speed > threshold).groupby("Time.year").sum()

result = xr.Dataset({"maximum": maximum_per_year, "n_exceeded": n_exceeded}, coords={"threshold": threshold})

# to look at the values:
computed = result.compute()
computed
```

**Reply by delahoussou** [reply]
I have been running it for about an hour an half now and still no output. I will paste what I get when it is ready.

**Reply by delahoussou** [reply]
![03FFC036-F8E9-4DE8-84AE-351DC5991FC0](https://user-images.githubusercontent.com/101612226/211094918-e1d76252-f5f1-46f4-b907-fc5fe3c42899.jpeg)

**Reply by keewis** [reply]
well, that can happen if your data is big (you can see that by displaying `result` and looking at the array's data repr). If you want to know more about `dask`, see [the dask documentation](https://docs.dask.org) and [this section from the xarray docs](https://docs.xarray.dev/en/stable/user-guide/dask.html).

**Comment by delahoussou** [comment]
![66FED0C5-6F63-4BB2-A778-F2778EB243E8](https://user-images.githubusercontent.com/101612226/211079373-9a6a77c3-421e-44d3-8052-166a4d33d4f8.jpeg)
